File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int n = grid.length, m = grid[0].length, cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    cnt++;
                    dfs(grid,i,j);
                }
            }
        }
        return cnt;
    }
    public void dfs(char[][] grid, int row, int col) {
        int n = grid.length, m = grid[0].length;
        if (row < 0 || col < 0 || row >= n || col >= m || grid[row][col] == '0') return;
        grid[row][col] = '0';
        dfs(grid, row + 1, col);
        dfs(grid, row - 1, col);
        dfs(grid, row, col + 1);
        dfs(grid, row, col - 1);
    }
}

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<char>>& edges, vector<vector<bool>>& vis) {
        int n = edges.size(), m = edges[0].size();
        if (row < 0 || col < 0 || row > n - 1|| col > m - 1 || vis[row][col] == true || edges[row][col] == '0') return;
        vis[row][col] = true;
        dfs(row + 1, col, edges, vis);
        dfs(row - 1, col, edges, vis);
        dfs(row, col + 1, edges, vis);
        dfs(row, col - 1, edges, vis);
    }
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size(), m = grid[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!vis[i][j] && grid[i][j] == '1') {
                    cnt++;
                    dfs(i, j, grid, vis);
                }
            }
        }
        return cnt;
    }
};